engine/execution/messages.go

```diff
@@ -52,3 +52,70 @@ func NewEmptyComputationResult(
 		},
 	}
 }
+
+func (cr *ComputationResult) CollectionResult(colIndex int) *ColResSnapshot {
+	if colIndex < 0 && colIndex > len(cr.CompleteCollections) {
+		return nil
+	}
+	return &ColResSnapshot{
+		BlockHeaderFunc: func() *flow.Header {
+			return cr.Block.Header
+		},
+		CollectionFunc: func() *flow.Collection {
+			return &flow.Collection{
+				Transactions: cr.CollectionAt(colIndex).Transactions,
+			}
+		},
+		UpdatedRegistersFunc: func() flow.RegisterEntries {
+			return cr.StateSnapshots[colIndex].UpdatedRegisters()
+		},
+		TouchedRegistersFunc: func() flow.RegisterIDs {
+			return cr.StateSnapshots[colIndex].AllRegisterIDs()
+		},
+		EmittedEventsFunc: func() flow.EventsList {
+			return cr.Events[colIndex]
+		},
+		TransactionResultsFunc: func() flow.TransactionResults {
+			var startTxnIndex int
+			if colIndex > 0 {
+				startTxnIndex = cr.TransactionResultIndex[colIndex-1]
+			}
+			endTxnIndex := cr.TransactionResultIndex[colIndex]
+			return cr.TransactionResults[startTxnIndex:endTxnIndex]
+		},
+	}
+
+}
+
+type ColResSnapshot struct {
+	BlockHeaderFunc        func() *flow.Header
```

Comment:
why not just implement them as methods?

Reply:
I had some thoughts initially but not valid anymore, so just changed it back
[066683c](https://github.com/onflow/flow-go/pull/4049/commits/066683c63c4de690855a8e6443d9a51c769e12dc)